### PR TITLE
Feature/164 orientation

### PR DESCRIPTION
**Pull request by ToufiPF**
Uses the compass instead of the gps to orient the user in the MapboxActivty.
I refactored a bit the previous code, creating dedicated interfaces for the compass and location sensor. 
CustomLocationProvider now has explicit dependencies to both a CompassSensor and LocationSensor.
To make movements more fluid, I allowed position and bearing camera updates to happen at the same time.

I have a problem with my phone's Wifi so I can't launch a tour on it. 
I made the video recording from the emulator, I don't know why it flickers like that.

https://user-images.githubusercontent.com/13187080/116108120-e8cc6c00-a6b3-11eb-9c57-1dbc8135349f.mp4

As always, all comments are appreciated !
Cheers

### CODE REVIEW

File: app/src/test/java/com/pindex/main/mapbox/orientation/OrientationSensorImplTest.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.pindex.main.mapbox.orientation
```
Comment (lucasrlt) [inline app/src/test/java/com/pindex/main/mapbox/orientation/OrientationSensorImplTest.kt:1]: Is this file needed in some way ? Or it should maybe be deleted

File: app/src/test/java/com/pindex/main/mapbox/orientation/MagnetometerOrientationSensorTest.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import android.hardware.SensorEventListener
+import android.hardware.SensorManager
+import com.nhaarman.mockitokotlin2.*
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mockito
+
+class MagnetometerOrientationSensorTest : AbstractOrientationSensorTest() {
+
+
```
Comment (lucasrlt) [inline app/src/test/java/com/pindex/main/mapbox/orientation/MagnetometerOrientationSensorTest.kt:13]: That's really nothing but there is an extra line break here :laughing:

**Review by lucasrlt — DISMISSED** [review]
Very impressive PR ! I didn't realize that much work was required for this feature but having to implement both sensors for different devices is really a lot 

+ incredible testing suite, well done ! 

Just left a comment about an empty file, but if it is needed for some reason I will approve the PR

File: app/src/main/java/com/pindex/main/mapbox/camera/CameraController.kt [hunk 3]
```diff
@@ -141,55 +148,82 @@ class CameraController(
...
+                val animatorSet = transitionEngine.linearTransition(camera)
+                updateCamera(animatorSet, POSITION_ANIMATION, options)
             }
             else -> { // either we're in transition or FREE state -> do nothing
             }
         }
     }
 
-    private fun startAnimation(animatorSet: AnimatorSet) {
+    private fun updateBearing(bearing: Double, options: (ValueAnimator.() -> Unit)?) {
```
[the remark this answers is not in the document]
Comment (ToufiPF) [reply app/src/main/java/com/pindex/main/mapbox/camera/CameraController.kt:169]: I don't think so, `ValueAnimator.() -> Unit` denotes a extension function (so it's basically an alias for `(ValueAnimator) -> Unit`)

File: app/src/test/java/com/pindex/main/mapbox/orientation/OrientationSensorImplTest.kt (hunk 1 above)
Comment (ToufiPF) [reply app/src/test/java/com/pindex/main/mapbox/orientation/OrientationSensorImplTest.kt:1]: Oops you're totally right MB

**Review by lucasrlt — APPROVED** [review]
Everything looks good to me, gj ! 👍

File: app/src/main/java/com/pindex/main/mapbox/camera/CameraController.kt [hunk 4]
```diff
@@ -141,55 +148,82 @@ class CameraController(
...
+        val camera = cameraOptions {
+            bearing(bearing)
+        }
+        println("Received bearing: $bearing°")
```
Comment (Mortadha-abderrhim) [inline app/src/main/java/com/pindex/main/mapbox/camera/CameraController.kt:173]: Should this print be kept?

File: app/src/main/java/com/pindex/main/mapbox/orientation/AbstractOrientationSensor.kt [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                )
+            ) remappedRotationMatrix else null
+    }
+
+    private fun notifyObservers(angles: FloatArray) {
+        val azimuth = Math.toDegrees(angles[0].toDouble())
+        val pitch = Math.toDegrees(angles[1].toDouble())
+        val roll = Math.toDegrees(angles[2].toDouble())
+
+        println("New angles: $azimuth°, $pitch°, $roll°")
```
Comment (Mortadha-abderrhim) [inline app/src/main/java/com/pindex/main/mapbox/orientation/AbstractOrientationSensor.kt:178]: Same here

File: app/src/main/java/com/pindex/main/mapbox/orientation/OrientationSensor.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+         * and a [MagnetometerOrientationSensor] otherwise.
+         */
+        @JvmStatic
+        fun getBestOrientationSensor(
+            context: Context,
+            updatesInterval: Long = DEFAULT_UPDATES_INTERVAL
+        ): OrientationSensor {
+            val sensorManager = ContextCompat.getSystemService(context, SensorManager::class.java)
+            if (CompassOrientationSensor.isSupported(sensorManager)) {
+                println("Compass is supported, using CompassOrientationSensor")
```
Comment (Mortadha-abderrhim) [inline app/src/main/java/com/pindex/main/mapbox/orientation/OrientationSensor.kt:30]: same in this class there are a few

**Review by Mortadha-abderrhim — APPROVED** [review]
Really Great Job!
I just think that these prints should be deleted (unless I am missing sth about what they actually do other than debug)
